### PR TITLE
OpenCL support for cracking AxCrypt 1.x

### PR DESCRIPTION
**Pull request by kholia**
I will rebase to remove the Python commit once https://github.com/magnumripper/JohnTheRipper/pull/3351 is merged.

This fixes https://github.com/magnumripper/JohnTheRipper/issues/3350 (Add OpenCL support for cracking AxCrypt).

### CODE REVIEW

File: src/opencl/axcrypt_kernel.cl [hunk 1]
```diff
@@ -0,0 +1,115 @@
...
+                (cp)[2] ^= (unsigned char)((value) >> 16), \
+                (cp)[3] ^= (unsigned char)((value) >> 24 ) )
+
+inline int axcrypt_decrypt(__global const axcrypt_password *inbuffer, uint gid, __constant axcrypt_salt *cur_salt, __global axcrypt_out *output)
+{
+	uchar password[PLAINTEXT_LENGTH];
+	uchar keyfile[4096];
+	int password_length = inbuffer[gid].length;
+	uchar salt[16];
+	unsigned char KEK[20], lsb[24], cipher[16];
```
Comment (magnumripper) [inline src/opencl/axcrypt_kernel.cl]: Why is lsb 24 bytes when you only ever touch the first 16?

File: src/opencl/axcrypt_kernel.cl [hunk 2]
```diff
@@ -0,0 +1,115 @@
...
+		/* AES_ECB(KEK, (MSB XOR (NUMBER_AES_BLOCKS * j + i)) | R[i]) */
+		AES_decrypt(cipher, cipher, &akey);
+		memcpy_macro(lsb + 8, cipher + 8, 8);
+
+		/* 2nd block treatment */
+		PUT_64BITS_XOR_MSB(cipher, 2 * j + 1);
+		memcpy_macro(cipher + 8, lsb, 8);
+		AES_decrypt(cipher, cipher, &akey);
+		memcpy_macro(lsb, cipher + 8, 8);
+	}
```
Comment (magnumripper) [inline src/opencl/axcrypt_kernel.cl]: This hot loop should probably be optimized - make `cipher` and `lsb` unions that also have 32-bit members, then replace all `memcpy_macro()` with 32-bit copies eg.
```diff
- 		memcpy_macro(lsb, cipher + 8, 8);
+ 		lsb.w[0] = cipher.w[2];
+ 		lsb.w[1] = cipher.w[3];
```
You could also write a 32-bit version of `PUT_64BITS_XOR_MSB()`

**Comment by kholia** [comment]
Hashcat seems to be lot faster,

```
Session..........: hashcat
Status...........: Cracked
Hash.Type........: AxCrypt
Hash.Target......: $axcrypt$*1*20000*f00875d0a137f83f2100fc8c9687ab25*...ca2ab0
Time.Started.....: Sat Aug  4 15:33:39 2018 (1 min, 16 secs)
Time.Estimated...: Sat Aug  4 15:34:55 2018 (0 secs)
Guess.Mask.......: ope?a?a?a?a?a [8]
Guess.Queue......: 1/1 (100.00%)
Speed.Dev.#7.....:    41754 H/s (480.31ms) @ Accel:128 Loops:128 Thr:1024 Vec:1
Recovered........: 1/1 (100.00%) Digests, 1/1 (100.00%) Salts
Progress.........: 3145728/7737809375 (0.04%)
Rejected.........: 0/3145728 (0.00%)
Restore.Point....: 0/7737809375 (0.00%)
Candidates.#7....: opeerane -> ope%uELL
HWMon.Dev.#7.....: Temp: 77c Fan: 34% Util: 99% Core:1151MHz Mem:3304MHz Bus:16
```

Our GPU utilization is pretty good but our speed is low, weird? I will remove the keyfile handling code (yes, we support that too), and test again.

File: src/opencl_axcrypt_fmt_plug.c [hunk 3]
```diff
@@ -0,0 +1,317 @@
...
+		"Copy data to gpu");
+
+	// Run kernel
+	BENCH_CLERROR(clEnqueueNDRangeKernel(queue[gpu_id], crypt_kernel, 1,
+		NULL, &gws, lws, 0, NULL,
+		multi_profilingEvent[1]),
+		"Run kernel");
+
+	// Read the result back
+	BENCH_CLERROR(clEnqueueReadBuffer(queue[gpu_id], mem_out, CL_TRUE, 0, outsize, output, 0, NULL, multi_profilingEvent[5]), "Copy result back");
```
Comment (magnumripper) [inline src/opencl_axcrypt_fmt_plug.c]: Bug here. This should be `multi_profilingEvent[2]`.

File: src/opencl/axcrypt_kernel.cl (hunk 2 above)
Comment (kholia) [reply src/opencl/axcrypt_kernel.cl]: Doing so might not be straightforward -> `cipher` and `lsb` need to hold different values before the loop starts.

File: src/opencl/axcrypt_kernel.cl (hunk 1 above)
Comment (kholia) [reply src/opencl/axcrypt_kernel.cl]: Ack. Fixing.

File: src/opencl/axcrypt_kernel.cl (hunk 2 above)
Comment (magnumripper) [reply src/opencl/axcrypt_kernel.cl]: Those operations can be left as-is (except for changing eg. `lsb` to `lsb.b`

**Comment by kholia** [comment]
@magnumripper Are you OK with the following patch?

```diff
diff --git a/src/opencl_axcrypt_fmt_plug.c b/src/opencl_axcrypt_fmt_plug.c
index a0ed013..b357157 100644
--- a/src/opencl_axcrypt_fmt_plug.c
+++ b/src/opencl_axcrypt_fmt_plug.c
@@ -159,7 +159,7 @@ static void reset(struct db_main *db)
                                       sizeof(axcrypt_password), 0, db);
 
                // Auto tune execution from shared/included code.
-               autotune_run(self, 1, 0, 200);
+               autotune_run(self, 1, 0, 800);
        }
 }
```

The speeds are much better with this tuning setting,

```
$ ../run/john --format=axcrypt-opencl h --mask='ope?a?a?a?a?a'
Device 6: GeForce GTX TITAN X
Using default input encoding: UTF-8
Loaded 1 password hash (axcrypt-opencl [SHA1 AES OpenCL])
Cost 1 (iteration count) is 20000 for all loaded hashes
-
Press 'q' or Ctrl-C to abort, almost any other key for status
0g 0:00:00:01 0.00% (ETA: 2018-08-08 08:01) 0g/s 18204p/s 18204c/s 18204C/s GPU:66C opeZ/1aa..opeisoaa
0g 0:00:00:02 0.00% (ETA: 2018-08-08 08:01) 0g/s 19125p/s 19125c/s 19125C/s GPU:67C opeISnaa..opeXQraa
0g 0:00:00:05 0.00% (ETA: 2018-08-08 17:45) 0g/s 19643p/s 19643c/s 19643C/s GPU:69C opembmaa..opeMJ3aa
0g 0:00:00:05 0.00% (ETA: 2018-08-08 17:45) 0g/s 19504p/s 19504c/s 19504C/s GPU:66C opembmaa..opeMJ3aa
```

**Comment by magnumripper** [comment]
A proper fix would be to make a split kernel, doing 1000 iterations per call or so. But 800 ms is tolerable I guess.

**Comment by kholia** [comment]
Ack. I haven't learned the split-kernel and gpu-side-mask stuff yet.

**Comment by magnumripper** [comment]
Split kernels are nothing to learn really, it's not harder than any other code you write. There's nothing to it, you just keep the "state" in global memory.

**Comment by magnumripper** [comment]
One thing you can try is `#define AES_NO_BITSLICE` before sourcing opencl_aes.h. Sometimes the other code is faster, sometimes not.

**Comment by magnumripper** [comment]
> One thing you can try is `#define AES_NO_BITSLICE` before sourcing opencl_aes.h

Oh and if you try this, be sure to drop the cache with `rm -rf ~/.nv/ComputeCache` (for nvidia), or it will run the cached binary and you'll see no change.

**Comment by kholia** [comment]
Using `#define AES_NO_BITSLICE` makes the code very slow in this case.

**Comment by kholia** [comment]
From https://www.traviscistatus.com/,

Investigating - We are seeing both private and open source builds routed to our sudo-enabled infrastructure (i.e. with `sudo: required`) that aren't starting at the moment. We are looking into it and we will provide an update as soon as possible. Thank you for your patience.

Aug 4, 12:06 UTC

**Comment by magnumripper** [comment]
Well it does pass CircleCI so I'll just merge it.

**Comment by magnumripper** [comment]
...or perhaps I shouldn't, until that other commit is merged! Do we have to wait for @Fist0urs ?

**Comment by kholia** [comment]
I will merge this tomorrow after rebasing, and also pushing a few tweaks.

Thanks!

**Comment by kholia** [comment]
> Do we have to wait for @Fist0urs ?

We should be good.

> I will merge this tomorrow after rebasing, and also pushing a few tweaks.

This is done now.

**Pull request merged**